In [1]:
import litellm
from dotenv import load_dotenv
from papersys.const import BASE_DIR

load_dotenv(BASE_DIR / ".env")


True

In [3]:
litellm.embedding

<function litellm.main.embedding(model, input=[], dimensions: Optional[int] = None, encoding_format: Optional[str] = None, timeout=600, api_base: Optional[str] = None, api_version: Optional[str] = None, api_key: Optional[str] = None, api_type: Optional[str] = None, caching: bool = False, user: Optional[str] = None, custom_llm_provider=None, litellm_call_id=None, logger_fn=None, **kwargs) -> Union[litellm.types.utils.EmbeddingResponse, Coroutine[Any, Any, litellm.types.utils.EmbeddingResponse]]>

In [5]:
litellm.model_cost['gemini-embedding-001']

{'input_cost_per_token': 1.5e-07,
 'litellm_provider': 'vertex_ai-embedding-models',
 'max_input_tokens': 2048,
 'max_tokens': 2048,
 'mode': 'embedding',
 'output_cost_per_token': 0,
 'output_vector_size': 3072,
 'source': 'https://cloud.google.com/vertex-ai/generative-ai/docs/learn/models'}

In [9]:
# 使用 litellm.embedding() 进行批量 embedding
response = litellm.embedding(
    model="gemini/gemini-embedding-001",  # 指定使用 Gemini embedding 模型
    input=["I love programming.", "Python is great for data science."]  # 批量输入文本
)

# 查看结果
print(f"模型: {response.model}")
print(f"嵌入维度: {len(response.data[0].embedding)}")
print(f"生成了 {len(response.data)} 个嵌入向量")
print(f"\n第一个文本的嵌入向量（前10维）: {response.data[0].embedding[:10]}")
print(f"第二个文本的嵌入向量（前10维）: {response.data[1].embedding[:10]}")

模型: gemini-embedding-001
嵌入维度: 3072
生成了 2 个嵌入向量

第一个文本的嵌入向量（前10维）: [-0.004167554, 0.010072836, 0.008006427, -0.085819386, 0.0007920959, 0.006507263, 0.019792652, 0.010993583, 0.006896781, 0.004768953]
第二个文本的嵌入向量（前10维）: [-0.0039723944, 0.0037366473, 0.010403045, -0.06857629, -0.009790573, -0.011662395, -0.003661297, 0.018717255, 0.015728533, -0.0025070773]


In [10]:
# 提取嵌入向量数据
embeddings = [item.embedding for item in response.data]

# 或者如果需要处理更多文本，可以这样批量处理
texts = [
    "Machine learning is fascinating.",
    "I enjoy coding in Python.",
    "Data science combines statistics and programming.",
    "AI is transforming the world."
]

batch_response = litellm.embehdding(
    model="gemini/gemini-embedding-001",
    input=texts
)

print(f"批量处理了 {len(batch_response.data)} 个文本")
print(f"每个嵌入向量的维度: {len(batch_response.data[0].embedding)}")

# 获取所有嵌入向量
all_embeddings = [item.embedding for item in batch_response.data]
print(f"嵌入向量列表形状: {len(all_embeddings)} x {len(all_embeddings[0])}")

批量处理了 4 个文本
每个嵌入向量的维度: 3072
嵌入向量列表形状: 4 x 3072


In [11]:
# 让我们先检查 litellm 中关于 batch embedding 的功能
import inspect

# 查看 google_batch_embeddings 模块
print("=== google_batch_embeddings 模块 ===")
print(dir(litellm.google_batch_embeddings))
print()

# 查看 batch_embeddings 函数的签名
print("=== batch_embeddings 函数签名 ===")
print(inspect.signature(litellm.google_batch_embeddings.batch_embeddings))
print()

# 查看函数文档
print("=== batch_embeddings 文档 ===")
print(litellm.google_batch_embeddings.batch_embeddings.__doc__)

=== google_batch_embeddings 模块 ===
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_check_custom_proxy', '_credentials', '_credentials_from_authorized_user', '_credentials_from_default_auth', '_credentials_from_identity_pool', '_credentials_from_identity_pool_with_aws', '_credentials_from_service_account', '_credentials_project_mapping', '_ensure_access_token', '_ensure_access_token_async', '_get_token_and_url', '_handle_reauthentication', 'access_token', 'async_batch_embeddings', 'async_completion', 'async_handler', 'async_streaming', 'batch_embeddings', 'completion', 'create_vertex_url', 'get_access_token', 'get_api_base', 'get_complete_vertex_url', 'get_default_vertex_locati

In [12]:
# 检查 litellm.embedding 是否支持 batch 参数
print("=== litellm.embedding 函数签名 ===")
print(inspect.signature(litellm.embedding))
print()

# 查看是否有其他 batch 相关的功能
print("=== litellm 中包含 'batch' 的属性 ===")
batch_attrs = [attr for attr in dir(litellm) if 'batch' in attr.lower()]
print(batch_attrs)

=== litellm.embedding 函数签名 ===
(model, input=[], dimensions: Optional[int] = None, encoding_format: Optional[str] = None, timeout=600, api_base: Optional[str] = None, api_version: Optional[str] = None, api_key: Optional[str] = None, api_type: Optional[str] = None, caching: bool = False, user: Optional[str] = None, custom_llm_provider=None, litellm_call_id=None, logger_fn=None, **kwargs) -> Union[litellm.types.utils.EmbeddingResponse, Coroutine[Any, Any, litellm.types.utils.EmbeddingResponse]]

=== litellm 中包含 'batch' 的属性 ===
['AzureBatchesAPI', 'Batch', 'BatchJobStatus', 'CancelBatchRequest', 'CreateBatchRequest', 'DEFAULT_BATCH_SIZE', 'GoogleBatchEmbeddings', 'ListBatchRequest', 'LiteLLMBatch', 'LiteLLMBatchCreateRequest', 'OpenAIBatchesAPI', 'RetrieveBatchRequest', 'VertexAIBatchPrediction', 'acancel_batch', 'acreate_batch', 'alist_batches', 'aretrieve_batch', 'argilla_batch_size', 'azure_batches_instance', 'batch_completion', 'batch_completion_models', 'batch_completion_models_all_r

In [13]:
# 查看 batch_completion 的签名和文档
print("=== batch_completion 函数 ===")
print(inspect.signature(litellm.batch_completion))
print()
print(litellm.batch_completion.__doc__)

=== batch_completion 函数 ===
(model: str, messages: List = [], functions: Optional[List] = None, function_call: Optional[str] = None, temperature: Optional[float] = None, top_p: Optional[float] = None, n: Optional[int] = None, stream: Optional[bool] = None, stop=None, max_tokens: Optional[int] = None, presence_penalty: Optional[float] = None, frequency_penalty: Optional[float] = None, logit_bias: Optional[dict] = None, user: Optional[str] = None, deployment_id=None, request_timeout: Optional[int] = None, timeout: Optional[int] = 600, max_workers: Optional[int] = 100, **kwargs)


    Batch litellm.completion function for a given model.

    Args:
        model (str): The model to use for generating completions.
        messages (List, optional): List of messages to use as input for generating completions. Defaults to [].
        functions (List, optional): List of functions to use as input for generating completions. Defaults to [].
        function_call (str, optional): The function cal

In [14]:
# 让我测试一下：当我们传递多个文本给 litellm.embedding 时，
# 它实际上调用的是什么 API

# 首先，让我们直接查看 Google AI 的 embedding API
# Google AI Studio 的 Gemini API 有两个端点：
# 1. embedContent - 单个文本
# 2. batchEmbedContents - 批量文本

# 让我们尝试使用 litellm 的详细日志来查看实际调用
litellm.set_verbose = True

print("测试调用 litellm.embedding 时的日志：")
test_response = litellm.embedding(
    model="gemini/gemini-embedding-001",
    input=["test 1", "test 2", "test 3"]
)

14:23:35 - LiteLLM:WARNING: utils.py:550 - `litellm.set_verbose` is deprecated. Please set `os.environ['LITELLM_LOG'] = 'DEBUG'` for debug logs.


测试调用 litellm.embedding 时的日志：
SYNC kwargs[caching]: False; litellm.cache: None; kwargs.get('cache')['no-cache']: False


In [2]:
# 关闭 verbose 模式
litellm.set_verbose = False

# 让我直接使用 Google 的 generativeai 库来测试 batch embedding
import google.generativeai as genai
import os

# 配置 API key
genai.configure(api_key=os.getenv("GEMINI_API_KEY"))

# 查看可用的方法
print("=== google.generativeai.embed_content 相关功能 ===")
print([attr for attr in dir(genai) if 'embed' in attr.lower()])
print()

# 测试 batch embedding
texts = ["text 1", "text 2", "text 3"]
result = genai.embed_content(
    model="models/embedding-001",
    content=texts,
    task_type="retrieval_document"
)

print(f"返回的嵌入数量: {len(result['embedding'])}")
print(f"第一个嵌入的维度: {len(result['embedding'][0]) if isinstance(result['embedding'][0], list) else 'scalar'}")
print(f"结果类型: {type(result)}")
print(f"结果键: {result.keys() if hasattr(result, 'keys') else 'not a dict'}")

=== google.generativeai.embed_content 相关功能 ===
['embed_content', 'embed_content_async']

返回的嵌入数量: 3
第一个嵌入的维度: 768
结果类型: <class 'dict'>
结果键: dict_keys(['embedding'])


In [3]:
# Google AI 的 embed_content 实际上支持批量输入
# 但让我检查一下是否有专门的 batch_embed_contents 方法

# 检查 genai 模块的所有方法
print("=== google.generativeai 所有公开方法 ===")
public_methods = [attr for attr in dir(genai) if not attr.startswith('_')]
for method in public_methods:
    if 'batch' in method.lower() or 'embed' in method.lower():
        print(f"  - {method}")
print()

# 实际上，根据 Google AI 的文档：
# embed_content 已经支持批量处理，当你传入列表时会自动批处理
# 但要使用 batch API（更便宜），需要使用 batch_embed_contents

# 让我查找是否有这个方法
try:
    print("检查 batch_embed_contents:")
    print(hasattr(genai, 'batch_embed_contents'))
    if hasattr(genai, 'batch_embed_contents'):
        print(genai.batch_embed_contents)
except Exception as e:
    print(f"错误: {e}")

=== google.generativeai 所有公开方法 ===
  - embed_content
  - embed_content_async

检查 batch_embed_contents:
False


## 使用 Batch Embedding API（价格减半！）

根据 Google 文档，Batch API 提供：
- **价格：正常价格的 50%** 💰
- **处理方式：异步处理**
- **返回时间：目标 24 小时内，通常更快**
- **适用场景：大量数据预处理、非紧急任务**

下面演示如何使用 Google 的新 SDK (`google-genai`) 来调用 Batch Embedding API：

In [4]:
# 方法1：使用 inline requests（适合少量请求）
from google import genai

client = genai.Client(api_key=os.getenv("GEMINI_API_KEY"))

# 准备要嵌入的文本列表
texts_to_embed = [
    "Machine learning is transforming technology.",
    "Python is excellent for data science.",
    "Natural language processing enables AI to understand text.",
    "Deep learning models require large datasets.",
    "Transfer learning reduces training time."
]

# 创建 inline embedding 请求
inline_requests = []
for idx, text in enumerate(texts_to_embed):
    request = {
        'content': text,
        'task_type': 'RETRIEVAL_DOCUMENT',  # 指定任务类型
    }
    inline_requests.append(request)

print(f"准备了 {len(inline_requests)} 个 embedding 请求")
print("创建 batch job...")

准备了 5 个 embedding 请求
创建 batch job...


In [6]:
# 让我先查看 batches.create_embeddings 的正确用法
import inspect

print("=== batches.create_embeddings 签名 ===")
print(inspect.signature(client.batches.create_embeddings))
print()

# 查看需要什么类型的数据
from google.genai import types
print("=== EmbeddingsBatchJobSource 的字段 ===")
print(types.EmbeddingsBatchJobSource.__annotations__)

=== batches.create_embeddings 签名 ===
(*, model: str, src: Union[google.genai.types.EmbeddingsBatchJobSource, google.genai.types.EmbeddingsBatchJobSourceDict], config: Union[google.genai.types.CreateEmbeddingsBatchJobConfig, google.genai.types.CreateEmbeddingsBatchJobConfigDict, NoneType] = None) -> google.genai.types.BatchJob

=== EmbeddingsBatchJobSource 的字段 ===
{'file_name': typing.Optional[str], 'inlined_requests': typing.Optional[google.genai.types.EmbedContentBatch]}


In [7]:
# 查看 EmbedContentBatch 的结构
print("=== EmbedContentBatch 的字段 ===")
print(types.EmbedContentBatch.__annotations__)
print()

# 查看 EmbedContentBatch 的文档
print("=== EmbedContentBatch 文档 ===")
print(types.EmbedContentBatch.__doc__)

=== EmbedContentBatch 的字段 ===
{'contents': typing.Union[google.genai.types.Content, str, PIL.Image.Image, google.genai.types.File, google.genai.types.Part, list[typing.Union[str, PIL.Image.Image, google.genai.types.File, google.genai.types.Part]], list[typing.Union[google.genai.types.Content, str, PIL.Image.Image, google.genai.types.File, google.genai.types.Part, list[typing.Union[str, PIL.Image.Image, google.genai.types.File, google.genai.types.Part]]]], NoneType], 'config': typing.Optional[google.genai.types.EmbedContentConfig]}

=== EmbedContentBatch 文档 ===
Parameters for the embed_content method.


In [8]:
# 正确的方式：使用 EmbedContentBatch
from google.genai import types

# 创建 batch embedding job
batch_job = client.batches.create_embeddings(
    model="gemini-embedding-001",
    src={
        'inlined_requests': types.EmbedContentBatch(
            contents=texts_to_embed,  # 直接传入文本列表
            config=types.EmbedContentConfig(
                task_type='RETRIEVAL_DOCUMENT',  # 指定任务类型
                output_dimensionality=768  # 可选：指定维度以节省存储空间
            )
        )
    },
    config={'display_name': "test-batch-embeddings"},
)

print(f"✅ Batch job 创建成功！")
print(f"Job name: {batch_job.name}")
print(f"Job state: {batch_job.state.name}")
print(f"\n💰 价格优势：Batch API 价格是正常价格的 50%！")
print(f"⏱️  处理时间：异步处理，通常几分钟到几小时内完成")

/tmp/ipykernel_17058/2424147294.py:5: ExperimentalWarning: batches.create_embeddings() is experimental and may change without notice.
  batch_job = client.batches.create_embeddings(


✅ Batch job 创建成功！
Job name: batches/ms9ssfwsac710gijdq6sft5jzz48jebfu50d
Job state: JOB_STATE_PENDING


AttributeError: 'BatchJob' object has no attribute 'config'

In [9]:
# 检查 job 状态
import time

job_name = batch_job.name
print(f"正在检查 job 状态: {job_name}\n")

# 轮询状态直到完成
completed_states = {
    'JOB_STATE_SUCCEEDED',
    'JOB_STATE_FAILED',
    'JOB_STATE_CANCELLED',
    'JOB_STATE_EXPIRED',
}

max_polls = 10  # 最多轮询 10 次（演示用）
poll_count = 0

while poll_count < max_polls:
    batch_job = client.batches.get(name=job_name)
    current_state = batch_job.state.name
    print(f"[{poll_count + 1}] 当前状态: {current_state}")
    
    if current_state in completed_states:
        print(f"\n🎉 Job 完成！最终状态: {current_state}")
        break
    
    poll_count += 1
    if poll_count < max_polls:
        print("   等待 5 秒后再次检查...")
        time.sleep(5)
else:
    print(f"\n⚠️  已轮询 {max_polls} 次，job 仍在处理中")
    print("   实际使用时可以保存 job_name，稍后再来获取结果")

正在检查 job 状态: batches/ms9ssfwsac710gijdq6sft5jzz48jebfu50d

[1] 当前状态: JOB_STATE_PENDING
   等待 5 秒后再次检查...
[2] 当前状态: JOB_STATE_SUCCEEDED

🎉 Job 完成！最终状态: JOB_STATE_SUCCEEDED


# 获取结果
if batch_job.state.name == 'JOB_STATE_SUCCEEDED':
    print("📦 获取 embedding 结果...\n")
    
    # 对于 inline requests，结果在 dest.inlined_embed_content_responses 中
    if batch_job.dest and batch_job.dest.inlined_embed_content_responses:
        responses = batch_job.dest.inlined_embed_content_responses
        print(f"✅ 成功获取 {len(responses)} 个 embedding！\n")
        
        # 先查看第一个响应的结构
        first_resp = responses[0]
        print(f"响应类型: {type(first_resp)}")
        print(f"响应对象: {first_resp.response}")
        print(f"响应内部类型: {type(first_resp.response)}")
        print(f"响应内部属性: {[attr for attr in dir(first_resp.response) if not attr.startswith('_')]}")
        print()
        
        # 检查 response.embedding
        if hasattr(first_resp.response, 'embedding'):
            print(f"embedding 属性: {type(first_resp.response.embedding)}")
            print(f"embedding 内容属性: {[attr for attr in dir(first_resp.response.embedding) if not attr.startswith('_')]}")
    else:
        print("⚠️  没有找到 inline responses")
else:
    print(f"❌ Job 未成功完成，状态: {batch_job.state.name}")

## 测试封装的 Batch Embedding 函数

现在测试我们封装好的 `google_batch_embedding` 函数，它使用 Google Batch API，价格减半！

In [13]:
# 重新导入以获取最新的代码
import importlib
import sys

# 如果模块已经导入，重新加载
if 'papersys.embedding' in sys.modules:
    importlib.reload(sys.modules['papersys.embedding'])

from papersys.embedding import google_batch_embedding

print("✅ 成功导入 google_batch_embedding 函数")

✅ 成功导入 google_batch_embedding 函数


In [14]:
# 准备测试数据
test_texts = [
    "机器学习正在改变世界",
    "Python 是数据科学的首选语言",
    "自然语言处理使 AI 能够理解文本",
    "深度学习需要大量的训练数据",
    "迁移学习可以减少训练时间",
    "神经网络模仿人脑的工作方式",
    "计算机视觉让机器能够看懂图像",
    "强化学习通过奖励来训练智能体"
]

print(f"准备嵌入 {len(test_texts)} 个中文文本")
print("\n文本列表:")
for i, text in enumerate(test_texts, 1):
    print(f"  {i}. {text}")

准备嵌入 8 个中文文本

文本列表:
  1. 机器学习正在改变世界
  2. Python 是数据科学的首选语言
  3. 自然语言处理使 AI 能够理解文本
  4. 深度学习需要大量的训练数据
  5. 迁移学习可以减少训练时间
  6. 神经网络模仿人脑的工作方式
  7. 计算机视觉让机器能够看懂图像
  8. 强化学习通过奖励来训练智能体


In [15]:
# 调用 google_batch_embedding 函数
import time

print("🚀 开始调用 Batch Embedding API...")
print("💰 提示：使用 Batch API 价格是正常 API 的 50%\n")

start_time = time.time()

# 调用函数，使用 768 维度以节省存储空间
embeddings_matrix = google_batch_embedding(
    model="gemini-embedding-001",
    inputs=test_texts,
    task_type="RETRIEVAL_DOCUMENT",
    output_dimensionality=768
)

elapsed_time = time.time() - start_time

print(f"✅ 成功！耗时: {elapsed_time:.2f} 秒\n")
print(f"返回的嵌入矩阵形状: {embeddings_matrix.shape}")
print(f"数据类型: {embeddings_matrix.dtype}")
print(f"每个向量的维度: {embeddings_matrix.shape[1]}")

🚀 开始调用 Batch Embedding API...
💰 提示：使用 Batch API 价格是正常 API 的 50%



/home/lyk/code/papersys/papersys/embedding.py:58: ExperimentalWarning: batches.create_embeddings() is experimental and may change without notice.
  batch_job = client.batches.create_embeddings(


✅ 成功！耗时: 61.87 秒

返回的嵌入矩阵形状: (8, 768)
数据类型: float32
每个向量的维度: 768


In [17]:
# 查看部分嵌入向量
import numpy as np

print("📊 查看前两个文本的嵌入向量（前10维）:\n")
for i in range(min(2, len(test_texts))):
    print(f"文本 {i+1}: {test_texts[i]}")
    print(f"Embedding: {embeddings_matrix[i][:10]}")
    print(f"向量范数: {np.linalg.norm(embeddings_matrix[i]):.4f}")
    print()

📊 查看前两个文本的嵌入向量（前10维）:

文本 1: 机器学习正在改变世界
Embedding: [-0.01820322  0.00460717  0.0146458  -0.08169461 -0.00463093  0.01815054
 -0.01603236  0.02238519  0.0202407   0.00680002]
向量范数: 0.5889

文本 2: Python 是数据科学的首选语言
Embedding: [-0.0145923   0.00761684  0.00508905 -0.08106402 -0.01248688  0.01986415
 -0.00407611  0.00463162  0.02045821 -0.00548646]
向量范数: 0.5856



In [18]:
# 计算文本之间的相似度
from sklearn.metrics.pairwise import cosine_similarity

# 归一化嵌入向量（对于 768 维度需要归一化）
normalized_embeddings = embeddings_matrix / np.linalg.norm(embeddings_matrix, axis=1, keepdims=True)

# 计算相似度矩阵
similarity_matrix = cosine_similarity(normalized_embeddings)

print("🔍 文本相似度分析:\n")
# 找出最相似的文本对
for i in range(len(test_texts)):
    for j in range(i + 1, len(test_texts)):
        similarity = similarity_matrix[i, j]
        if similarity > 0.8:  # 只显示高度相似的
            print(f"相似度 {similarity:.4f}:")
            print(f"  文本1: {test_texts[i]}")
            print(f"  文本2: {test_texts[j]}")
            print()

🔍 文本相似度分析:

相似度 0.8102:
  文本1: 机器学习正在改变世界
  文本2: Python 是数据科学的首选语言

相似度 0.8388:
  文本1: 机器学习正在改变世界
  文本2: 自然语言处理使 AI 能够理解文本

相似度 0.8366:
  文本1: 机器学习正在改变世界
  文本2: 深度学习需要大量的训练数据

相似度 0.8143:
  文本1: 机器学习正在改变世界
  文本2: 迁移学习可以减少训练时间

相似度 0.8074:
  文本1: 机器学习正在改变世界
  文本2: 神经网络模仿人脑的工作方式

相似度 0.8608:
  文本1: 机器学习正在改变世界
  文本2: 计算机视觉让机器能够看懂图像

相似度 0.8250:
  文本1: 机器学习正在改变世界
  文本2: 强化学习通过奖励来训练智能体

相似度 0.8219:
  文本1: 自然语言处理使 AI 能够理解文本
  文本2: 神经网络模仿人脑的工作方式

相似度 0.8551:
  文本1: 自然语言处理使 AI 能够理解文本
  文本2: 计算机视觉让机器能够看懂图像

相似度 0.8035:
  文本1: 自然语言处理使 AI 能够理解文本
  文本2: 强化学习通过奖励来训练智能体

相似度 0.8371:
  文本1: 深度学习需要大量的训练数据
  文本2: 迁移学习可以减少训练时间

相似度 0.8007:
  文本1: 深度学习需要大量的训练数据
  文本2: 神经网络模仿人脑的工作方式

相似度 0.8264:
  文本1: 深度学习需要大量的训练数据
  文本2: 计算机视觉让机器能够看懂图像

相似度 0.8085:
  文本1: 深度学习需要大量的训练数据
  文本2: 强化学习通过奖励来训练智能体

相似度 0.8177:
  文本1: 神经网络模仿人脑的工作方式
  文本2: 计算机视觉让机器能够看懂图像

相似度 0.8098:
  文本1: 计算机视觉让机器能够看懂图像
  文本2: 强化学习通过奖励来训练智能体



In [19]:
# 演示查询功能
query_text = "AI 和机器学习的关系"
print(f"🔎 查询: '{query_text}'\n")

# 获取查询的嵌入向量（使用 RETRIEVAL_QUERY 任务类型）
query_embedding = google_batch_embedding(
    model="gemini-embedding-001",
    inputs=[query_text],
    task_type="RETRIEVAL_QUERY",  # 查询使用不同的任务类型
    output_dimensionality=768
)

# 归一化查询向量
query_embedding_norm = query_embedding / np.linalg.norm(query_embedding)

# 计算与所有文档的相似度
similarities = cosine_similarity(query_embedding_norm, normalized_embeddings)[0]

# 排序并显示最相关的文档
top_k = 3
top_indices = np.argsort(similarities)[::-1][:top_k]

print(f"最相关的 {top_k} 个文档:\n")
for rank, idx in enumerate(top_indices, 1):
    print(f"{rank}. [相似度: {similarities[idx]:.4f}] {test_texts[idx]}")

🔎 查询: 'AI 和机器学习的关系'



/home/lyk/code/papersys/papersys/embedding.py:58: ExperimentalWarning: batches.create_embeddings() is experimental and may change without notice.
  batch_job = client.batches.create_embeddings(


最相关的 3 个文档:

1. [相似度: 0.7143] 机器学习正在改变世界
2. [相似度: 0.6679] 自然语言处理使 AI 能够理解文本
3. [相似度: 0.6455] 深度学习需要大量的训练数据


## 总结

### ✅ 成功实现的功能

1. **封装了 `google_batch_embedding` 函数**
   - 位置：`papersys/embedding.py`
   - 使用 Google Gemini Batch Embedding API
   - 支持自定义任务类型和输出维度
   - 自动处理异步等待和错误处理

2. **关键优势**
   - 💰 **价格减半**：Batch API 价格是正常 API 的 50%
   - 🚀 **批量处理**：一次性处理多个文本
   - 🎯 **任务优化**：支持多种任务类型（检索、分类、聚类等）
   - 📦 **便捷使用**：返回标准的 numpy 数组，易于集成

3. **测试结果**
   - ✅ 成功嵌入 8 个中文文本
   - ✅ 生成 768 维嵌入向量
   - ✅ 计算文本相似度正常
   - ✅ 语义搜索功能正常

### 📝 使用示例

```python
from papersys.embedding import google_batch_embedding

# 获取文档嵌入
embeddings = google_batch_embedding(
    model="gemini-embedding-001",
    inputs=["文本1", "文本2", "文本3"],
    task_type="RETRIEVAL_DOCUMENT",  # 文档检索
    output_dimensionality=768  # 可选：768, 1536, 3072
)

# 获取查询嵌入
query_embedding = google_batch_embedding(
    model="gemini-embedding-001",
    inputs=["查询文本"],
    task_type="RETRIEVAL_QUERY",  # 查询检索
    output_dimensionality=768
)
```

### 💡 建议

1. **维度选择**：使用 768 维可以节省 75% 的存储空间，性能损失很小
2. **任务类型**：文档使用 `RETRIEVAL_DOCUMENT`，查询使用 `RETRIEVAL_QUERY`
3. **归一化**：对于非 3072 维的嵌入，需要手动归一化后再计算相似度
4. **批量大小**：单次可处理大量文本，建议根据实际情况分批

---

## 附录：API 对比

| 特性 | 普通 Embedding API | Batch Embedding API |
|------|-------------------|---------------------|
| **价格** | 标准价格 | **50% 折扣** 💰 |
| **处理方式** | 同步，立即返回 | 异步，等待完成 |
| **响应时间** | 毫秒级 | 几秒到几分钟 |
| **适用场景** | 实时查询 | 批量预处理 |
| **最大批量** | 有限制 | 支持大量文本 |
| **推荐用途** | 在线搜索、实时应用 | 数据库构建、离线处理 |

### 何时使用 Batch API？

✅ **适合使用 Batch API**：
- 构建向量数据库
- 预处理大量文档
- 离线数据分析
- 成本敏感的应用

❌ **不适合使用 Batch API**：
- 实时用户查询
- 需要立即响应的场景
- 单个文本的嵌入